### PR TITLE
[modbus] Modbus transport API simplification

### PR DESCRIPTION
**Pull request by ssalonen**
Co-authored with @mrbig, user feedback from @Rossko57 .

As discussed in https://github.com/openhab/openhab-core/issues/1435#issuecomment-616706285.

This is a rather large PR, with the aim of improving the modbus transport API developer experience based on feedback from @mrbig and sunspec bundle.

The main changes introduced:
1. callbacks are `@Functional`, allowing binding developer to use lambdas and much less verbose way to access the data read. The API is vaguely modeled after vertx AsyncResult: https://vertx.io/docs/apidocs/io/vertx/core/AsyncResult.html

```java
comms.submitOneTimePoll(request, result -> {
            if (result.hasError()) {
                Exception error = (@NonNull Exception) result.getCause();
                handleError(error);
            } else {
                ModbusRegisterArray registers = (@NonNull ModbusRegisterArray) result.getRegisters();
                modelBlockReceived(registers);        
            });
```

Previously one had to have callback with three methods, but depending on the type of request, only two methods are actually applicable.
```java
BasicModbusReadRequestBlueprint request = new BasicModbusReadRequestBlueprint(slaveId,
        ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, baseAddress, // Start address
        count,
        maxTries);
PollTask task = new BasicPollTaskImpl(endpoint, request, new ModbusReadCallback() {

    @Override
    public void onRegisters(ModbusReadRequestBlueprint request, ModbusRegisterArray registers) {
        handleReceived(registers);
    }

    @Override
    public void onError(ModbusReadRequestBlueprint request, Exception error) {
        handleError(error);             
    }

    @Override
    public void onBits(@Nullable ModbusReadRequestBlueprint request, @Nullable BitArray bits) {
        // this is never called by transport bundle since request is for reading registers, not "bits" 
    }
});
handler.getManagerRef().get().submitOneTimePoll(task);
```
2. The other change is that transport API calls take in relevant parameters directly, and it's not necessary to "bundle" arguments into a `BasicPollTaskImpl `. See example above
3. There were many java interfaces which had only one implementation. We have removed the interfaces and use the implementation class in many places. There were also trivial implementations which did not really add value (e.g. specialized bit array implementation for cases with single bit) I think this reduces the cognitive load when getting familiar with the transport bundle. Commits can be recognized with "abstractions removed"

The above changes are not introducing any changes in behaviour.

-------------------------------------------------------------------

There is also an API change which is changing slightly how to interact with transport API: introduction of ModbusCommunicationInterface. 

Now, the transport API consumers need to "open" a ModbusCommunicationInterface to interact with a particular Modbus server ("slaves"). Modbus slaves are identified by `tcphost:tcpport` or `serialport`. Once you have been finished interacting with the slave, you `close()` the communication interface. There can be many communication interfaces open to the same slave.

Closing the communication interface signals to transport bundle that you are not expecting any further requests for a while. The transport bundle automatically frees resources (e.g. disconnects TCP connections) immediately if the last communication interface pointing to a server is closed. 

The eager cleanup of resources was one of the main motivations for this change. In addition, some of the API details became simpler as well (e.g. error handling when there are conflicting baud rates trying to be associated with the same serial port).

In practice, the communication interface is held by the thing representing the modbus slave ("tcp" or "serial" thing). Once the thing is disposed, also the communication interface is closed.

-------------------------------------------------------------------

Third change is improving automatic disconnection of unused idle connections. Previously connection age was checked after every request, and disconnected if older than the configured value. This had the downside that connection pool can contain open old connections when there are no transactions.

Now, the connection pool is inspected every 10 seconds, and old idle connections are disconnected.

For the user this is quite a transparent change, and should just mean that openHAB is not keeping any connections for long periods of time.

### CODE REVIEW

**Comment by ssalonen** [comment]
Prebuilt jars for commit 9a65a4abe2e9845b351e5adcc569f6a2bb50e2c7 (June 23, 2020)
[snapshot-commit-9a65a4abe2e9845b351e5adcc569f6a2bb50e2c7.zip](https://github.com/openhab/openhab-addons/files/4821054/snapshot-commit-9a65a4abe2e9845b351e5adcc569f6a2bb50e2c7.zip)

File: bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java [hunk 1]
```diff
@@ -280,30 +256,20 @@ private void modelBlockReceived(ModbusRegisterArray registers) {
...
-            @Override
-            public void onRegisters(ModbusReadRequestBlueprint request, ModbusRegisterArray registers) {
-                parseCommonBlock(registers);
-            }
-
-            @Override
-            public void onError(ModbusReadRequestBlueprint request, Exception error) {
+        comms.submitOneTimePoll(request, result -> {
+            if (result.hasError()) {
+                Exception error = (@NonNull Exception) result.getCause();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java:266]: Why not just have `getCause()` return an optional so you can use it to replace `hasError` while also avoiding all of the `@NonNull` casting you are doing?

File: bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java [hunk 2]
```diff
@@ -158,30 +154,20 @@ public void detectModel() {
...
                 handleError(error);
-            }
-
-            @Override
-            public void onBits(@Nullable ModbusReadRequestBlueprint request, @Nullable BitArray bits) {
-                // don't care, we don't expect this result
+            } else {
+                ModbusRegisterArray registers = (@NonNull ModbusRegisterArray) result.getRegisters();
+                headerReceived(registers);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java:168]: same suggestion here

**Comment by fwolter** [comment]
> ```java
> comms.submitOneTimePoll(request, result -> {
>             if (result.hasError()) {
>                 Exception error = (@NonNull Exception) result.getCause();
>                 handleError(error);
>             } else {
>                 ModbusRegisterArray registers = (@NonNull ModbusRegisterArray) result.getRegisters();
>                 modelBlockReceived(registers);        
>             });
> ```

I think error handling is mandatory. Why not adding a second functional interface to `submitOneTimePoll()` handling the error case?

`comms.submitOneTimePoll(request, this::handleError, result -> {`
...

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java [hunk 3]
```diff
@@ -652,7 +653,36 @@ private boolean containsOpenClosed(List<Class<? extends State>> acceptedDataType
     }
 
     @Override
-    public synchronized void onRegisters(ModbusReadRequestBlueprint request, ModbusRegisterArray registers) {
+    public synchronized void handle(AsyncModbusReadResult result) {
+        if (result.hasError()) {
+            Exception error = result.getCause();
+            assert error != null;
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java:659]: Assert here is guaranteed to work or the API contract is violated, it is there for the developer.

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java [hunk 4]
```diff
@@ -72,66 +63,33 @@
...
-                if (lastError == null) {
-                    this.lastError = new AtomicStampedKeyValue<>(System.currentTimeMillis(), request, error);
-                } else {
-                    lastError.update(System.currentTimeMillis(), request, error);
-                }
+            logger.debug("Thing {} received response {}", thing.getUID(), result);
+            childCallbacks.forEach(handler -> handler.handle(result));
+            if (result.hasError()) {
+                Exception error = result.getCause();
+                assert error != null;
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java:87]: Assert here is guaranteed to work or the API contract is violated, it is there for the developer.

File: bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java (hunk 1 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java:266]: Makes sense actually...👍

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusReadRequestBlueprint.java [hunk 5]
```diff
@@ -12,47 +12,121 @@
...
+ * if they have
+ * the equal parameters (same slave id, start, length, function code and maxTries).
  *
  * @author Sami Salonen - Initial contribution
  *
  */
 @NonNullByDefault
-public interface ModbusReadRequestBlueprint extends ModbusRequestBlueprint {
+public class ModbusReadRequestBlueprint {
+    private static StandardToStringStyle toStringStyle = new StandardToStringStyle();
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusReadRequestBlueprint.java:36]: This PR is not introducing it though.

I am happy to make this change but perhaps we should move it to separate PR to get this big one going?

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusRegisterArray.java [hunk 6]
```diff
@@ -70,7 +119,7 @@ default String toHexString() {
      * Appends the register data as hex string to the given StringBuffer
      *
      */
-    default StringBuffer appendHexString(StringBuffer buffer) {
+    public StringBuffer appendHexString(StringBuffer buffer) {
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusRegisterArray.java:122]: Same comment as above, would prefer this PR to move and introduce these general fixes separately

**Comment by ssalonen** [comment]
@fwolter very valid suggestion! With @mrbig we tried to model other similar APIS which bundle errors and valid results into same object

We played with several different API designs but to be honest, separating error from the successful result was not one of them.

By separating the error case the API consumer can avoid an if statement. On the other hand, bundling error case to the same result object makes it clear when the operation has ended (either successfully or with error). Are there any other upsides/downsides?

Either way will certainly work. Any thoughts @mrbig ?

I am not too keen on changing this if the gained benefit is small - it is somewhat large change considering the large footprint of the tests.

**Comment by fwolter** [comment]
Handling result and exception paths separately reflects OOP paradigm. You don't need the casting and the `getCause()` invocation. Furthermore, the compiler complains, if the user forgets to handle the error.

I can't follow your argument. not being clear when the operation finished. You could specify either the error handler *or* the result handler is invoked, when the operation finished.

**Comment by ssalonen** [comment]
@fwolter I like your thinking. I think it's especially valuable that you have to handle the errors and decide what to do with them.

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusReadRequestBlueprint.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/ModbusReadRequestBlueprint.java:36]: fine by me

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java [hunk 7]
```diff
@@ -736,123 +740,170 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
...
+            Objects.requireNonNull(localScheduledThreadPoolExecutor, "Not activated!");
+            WriteTask task = new BasicWriteTask(endpoint, request, callback);
+            long scheduleTime = System.currentTimeMillis();
+            logger.debug("Scheduling one-off write task {}", task);
+            ScheduledFuture<?> future = localScheduledThreadPoolExecutor.schedule(() -> {
+                long millisInThreadPoolWaiting = System.currentTimeMillis() - scheduleTime;
+                logger.debug("Will now execute one-off write task {}, waited in thread pool for {}", task,
+                        millisInThreadPoolWaiting);
+                executeOperation(task, true, writeOperation);
+            }, 0L, TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java:865]: Coming in via https://github.com/mrbig/openhab2-addons/pull/1/commits/8eb186d81ea83b1ccac56c8b1acc41778be2306e

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java [hunk 8]
```diff
@@ -736,123 +740,170 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
...
+                @Nullable
+                ScheduledFuture<?> future = scheduledPollTasks.remove(task);
+                if (future == null) {
+                    // No such poll task
+                    logger.warn("Caller tried to unregister nonexisting poll task {}", task);
+                    return false;
+                }
+                logger.info("Unregistering regular poll task {} (interrupting if necessary)", task);
+                future.cancel(true);
+                logger.info("Poll task {} canceled", task);
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java:844]: Coming in via https://github.com/mrbig/openhab2-addons/pull/1/commits/e61a970366e48dbf75bad8002e432f1a4f1aa21a

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java [hunk 9]
```diff
@@ -736,123 +740,170 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
...
-            Objects.requireNonNull(factory, "Not activated!");
+            long scheduleTime = System.currentTimeMillis();
+            BasicPollTask task = new BasicPollTask(endpoint, request, callback);
+            logger.debug("Scheduling one-off poll task {}", task);
+            ScheduledFuture<?> future = executor.schedule(() -> {
+                long millisInThreadPoolWaiting = System.currentTimeMillis() - scheduleTime;
+                logger.debug("Will now execute one-off poll task {}, waited in thread pool for {}", task,
+                        millisInThreadPoolWaiting);
+                executeOperation(task, true, pollOperation);
+            }, 0L, TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java:774]: Coming in via https://github.com/mrbig/openhab2-addons/pull/1/commits/67f8ecab9de26745496a7d0b31f2e960be0e4ccf

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/AbstractModbusEndpointThingHandler.java [hunk 10]
```diff
@@ -92,33 +96,26 @@ public void initialize() {
...
-    @Override
-    public @Nullable ModbusSlaveEndpoint asSlaveEndpoint() {
-        return endpoint;
+        try {
+            ModbusCommunicationInterface localComms = comms;
+            if (localComms != null) {
+                localComms.close();
+            }
+        } catch (Exception e) {
+            logger.error("Error closing modbus communication interface", e);
```
[the remark this answers is not in the document]
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/AbstractModbusEndpointThingHandler.java:105]: Coming in via https://github.com/mrbig/openhab2-addons/pull/1/commits/55997482b5c54733d1d825269e682be5f6a3b319

**Comment by ssalonen** [comment]
Mrbig & me worked on this more and now errors are propagated via their own callbacks. We also use Optionals to wrap the data in callbacks

File: bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java [hunk 11]
```diff
@@ -739,136 +743,177 @@ private void verifyTaskIsRegistered(PollTask task) throws PollTaskUnregistered {
...
+                    logger.trace("Unregistering previous poll task (possibly with different period)");
+                    unregisterRegularPoll(task);
+                }
+                ScheduledFuture<?> future = executor.scheduleWithFixedDelay(() -> {
+                    long started = System.currentTimeMillis();
+                    logger.debug("Executing scheduled ({}ms) poll task {}. Current millis: {}", pollPeriodMillis, task,
+                            started);
+                    try {
+                        executeOperation(task, false, pollOperation);
+                    } catch (RuntimeException e) {
```
Comment (ssalonen) [inline bundles/org.openhab.io.transport.modbus/src/main/java/org/openhab/io/transport/modbus/internal/ModbusManagerImpl.java:787]: Cherry-picked 5bde9d6604375f1212789dae5e002c5b7e9ceff3 on rebase in this commit.

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java [hunk 12]
```diff
@@ -805,7 +820,13 @@ public synchronized void onWriteResponse(ModbusWriteRequestBlueprint request, Mo
      * @return updated channel data
      */
     private Map<ChannelUID, State> processUpdatedValue(State numericState, boolean boolValue) {
-        Map<@NonNull ChannelUID, @NonNull State> states = new HashMap<>();
+        Transformation localReadTransformation = readTransformation;
+        if (localReadTransformation == null) {
+            // We should always have transformation available if thing is initalized properly
+            logger.trace("No transformation available, aborting processUpdatedValue");
+            return new HashMap<ChannelUID, State>();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java:827]: Could you return a `Collection.emptyMap()` here?

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java [hunk 13]
```diff
@@ -246,21 +177,48 @@ public ModbusPollerReadRequest(ModbusPollerConfiguration config,
...
 
-    private final Logger logger = LoggerFactory.getLogger(ModbusPollerThingHandlerImpl.class);
+    /**
+     * Immutable data object to cache the results of a poll request
+     */
+    private class PollResult {
+        @Nullable
+        public final AsyncModbusReadResult result;
+        @Nullable
+        public final AsyncModbusFailure<ModbusReadRequestBlueprint> failure;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java:187]:
```suggestion
        public final @Nullable AsyncModbusReadResult result;
        public final @Nullable AsyncModbusFailure<ModbusReadRequestBlueprint> failure;
```

Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusPollerThingHandler.java]: Resolved this, and many others in https://github.com/openhab/openhab-addons/pull/7994/commits/00b7f764be87d957e6b81354b54d7bdeb33cb818

File: bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java (hunk 12 above)
Comment (ssalonen) [reply bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/handler/ModbusDataThingHandler.java]: Yup. https://github.com/openhab/openhab-addons/pull/7994/commits/6a46de1e2f5329b0839c2aaaf1cceb55da22dd38

**Pull request merged**